src/merkle/new_smt/mod.rs

```diff
@@ -0,0 +1,272 @@
+use core::cmp::Ordering;
+
+use winter_math::StarkField;
+
+use crate::hash::rpo::Rpo256;
+use crate::utils::{collections::Vec, vec};
+use crate::{Felt, EMPTY_WORD};
+
+use super::{
+    BTreeMap, BTreeSet, EmptySubtreeRoots, InnerNode, LeafIndex, MerkleError, NodeIndex, RpoDigest,
+    SparseMerkleTree, Word,
+};
+
+#[cfg(test)]
+mod tests;
+
+// CONSTANTS
+// ================================================================================================
+
+pub const NEW_SMT_DEPTH: u8 = 64;
+
+// SMT
+// ================================================================================================
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct NewSmt {
+    root: RpoDigest,
+    leaves: BTreeMap<u64, NewSmtLeaf>,
+    inner_nodes: BTreeMap<NodeIndex, InnerNode>,
+}
+
+impl NewSmt {
+    /// Returns a new [NewSmt].
+    ///
+    /// All leaves in the returned tree are set to [ZERO; 4].
+    pub fn new() -> Self {
+        let root = *EmptySubtreeRoots::entry(NEW_SMT_DEPTH, 0);
+
+        Self {
+            root,
+            leaves: BTreeMap::new(),
+            inner_nodes: BTreeMap::new(),
+        }
+    }
+
+    /// Returns a new [SimpleSmt] instantiated with leaves set as specified by the provided entries.
+    ///
+    /// All leaves omitted from the entries list are set to [ZERO; 4].
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The number of entries exceeds 2^63 entries.
+    /// - The provided entries contain multiple values for the same key.
+    pub fn with_entries(
+        entries: impl IntoIterator<Item = (NewSmtKey, Word)>,
+    ) -> Result<Self, MerkleError> {
+        // create an empty tree
+        let mut tree = Self::new();
+
+        // compute the max number of entries. We use an upper bound of depth 63 because we consider
+        // passing in a vector of size 2^64 infeasible.
+        let max_num_entries = 2_usize.pow(tree.depth().min(63).into());
```

Comment:
Imported this from `SimpleSmt`; do we really need it?

Reply:
Yep - I don't think this is needed here.